### DISCUSSION TITLE
Using Word Level Lexicon to Decrease WER of ASR Models

### DISCUSSION THREAD

**Discussion by ahkarami** [NVIDIA-NeMo/NeMo#4094] [Q&A] [answered]
@titu1994  &  @VahidooX 
Hi,
I have 2 questions:
**1-** We know that, NeMo provides lexicon-free decoding with _e.g._, BPE LM with great results (for models like CitriNet), and it is obviously clear that it is a great advantage. However, imagine that we have already a great word level lexicon and we want to decrease the WER of ASR models by using it in a specific domain (without fine tuning our general ASR model on it (_i.e._, we don't want the cost of data annotation to be imposed on us)). How we can do that? I mean that is there any solution in NeMo to do that (or in future NeMo team wants to provide solutions for using lexicon)? 
Following the above question, I recently founded the [great library](https://github.com/kensho-technologies/pyctcdecode) (thanks @VahidooX for introducing it in [this discussion](https://github.com/NVIDIA/NeMo/discussions/4064)) which we can pass a `hotwords` list to its decoder. I haven't tested it yet, but I think we can't pass our whole lexicon as a `hotwords` list to it. According to your experiences, do you think that using it can decrease WER of ASR models in specific domains? 
**2-** My second question is that, how we can use a raw text corpus to create LM (without any audio files) to improve the final WER of an ASR model in NeMo? I know that your team provided an [excellent doc](https://docs.nvidia.com/deeplearning/nemo/user-guide/docs/en/stable/asr/asr_language_modeling.html#train-n-gram-lm) for it. But my exact question is that: if we, for example, crawled raw texts about a specific domain and put each sentence in a separate line of a text file then can we create an n-gram BPE based LM for it? If yes, is there any possibility to create a **word-level LM** in NeMo (or other libraries like `pyctcdecode`)? 
Best

**Comment by VahidooX** [comment, marked as answer]
1-We are doing research on providing robust and efficient solutions for such use cases but they are still in experimental stages. The easiest solution is to use word boosting (hotwords) with the beam search decoder as you mentioned. The second option can be to use a pretrained TTS model to create synthetic speech and fine-tune your ASR model on the new data. Third one is to use LM trained on the specific domain.

2-yes, you can train an LM on text data from your domain to adopt your ASR model to a specific domain. You can also merge your domain-specific text data with texts from general-domains (you can use the text corpus of LS which has around 4M sentences for this purpose). Our default beam search decoder only supports sub-word level LM while pyctcdecode supports word-level LM. You can use either of those. I recommend to start with our docs and our default decoder. You also have the option to train a neural rescorer on your own data. Using both the N-gram LM and neural rescorer is also an option here. One of them or both can be trained on your own domain-specific data.

**Reply by ahkarami** [reply]
@VahidooX 
Thanks for your great and complete answer.
Best
